### PR TITLE
feat: add common withdraw extension with lockability

### PR DESCRIPTION
**Pull request by kasrakhosravi**
## Description:

* Merging `ERC721SimpleProceeds` with `EmergencyOwnerWithdraw` extensions
* Combining `withdraw` methods into a single one with ability to set `amounts` to be withdrawn
* Test cases

## PR Checklist
- [x] :eyes: I have self-reviewed my PR, leaving comments on any changes which may trigger a discussion
- [x] :pencil2: Added relevant tests

### CODE REVIEW

File: contracts/common/WithdrawExtension.sol [hunk 1]
```diff
@@ -45,11 +68,32 @@ abstract contract EmergencyOwnerWithdrawExtension is
...
+        _registerInterface(type(IWithdrawExtension).interfaceId);
     }
 
     /* ADMIN */
 
+    function setProceedsRecipient(address _proceedsRecipient)
+        external
+        onlyOwner
+    {
+        require(!proceedsRecipientLocked, "Common/RECIPIENT_LOCKED");
```
Comment (aramalipoor) [inline contracts/common/WithdrawExtension.sol:80]:
```suggestion
        require(!proceedsRecipientLocked, "WITHDRAW/RECIPIENT_LOCKED");
```

File: contracts/common/WithdrawExtension.sol [hunk 2]
```diff
@@ -12,14 +12,20 @@ import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
...
 
-interface IEmergencyOwnerWithdrawExtension {
+interface IWithdrawExtension {
+    function recipientWithdraw() external;
+
     function emergencyWithdraw(address[] calldata claimTokens) external;
 
+    function setProceedsRecipient(address _proceedsRecipient) external;
+
+    function lockProceedsRecipient() external;
```
Comment (aramalipoor) [inline contracts/common/WithdrawExtension.sol:22]: Probably better to use more generic name, like `setWithdrawRecipient` and `lockWithdrawRecipient`

File: contracts/streams/ERC721/presets/ERC721EqualEmissionStream.sol [hunk 3]
```diff
@@ -20,7 +20,7 @@ contract ERC721EqualEmissionStream is
     Ownable,
     ERC721EmissionReleaseExtension,
     ERC721EqualSplitExtension,
-    EmergencyOwnerWithdrawExtension
+    WithdrawExtension
```
Comment (aramalipoor) [inline contracts/streams/ERC721/presets/ERC721EqualEmissionStream.sol:23]: Let's also update the ERC721 presets to use this extension and remove the "proceeds" extension from collections 🚀

File: contracts/common/WithdrawExtension.sol [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        __EmergencyOwnerWithdrawExtension_init_unchained();
+    }
+
+    function __EmergencyOwnerWithdrawExtension_init_unchained()
+        internal
+        onlyInitializing
+    {
+        _registerInterface(type(IWithdrawExtension).interfaceId);
+    }
```
Comment (aramalipoor) [inline contracts/common/WithdrawExtension.sol:74]: It's repeated, can be removed

Comment (kasrakhosravi) [reply contracts/common/WithdrawExtension.sol:75]: you mean remove `__EmergencyOwnerWithdrawExtension_init()` and replace it with `__WithdrawExtension_init`? In that case the recipient has to be passed to the `__WithdrawExtension_init`

Comment (aramalipoor) [reply contracts/common/WithdrawExtension.sol:75]: Yep that's correct, we only need one initializer, with same name as contract

File: contracts/common/WithdrawExtension.sol [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    function _assertWithdrawAccess(WithdrawMode withdrawMode, address account) internal returns (bool) {
+        if (withdrawMode == WithdrawMode.NOBODY) {
+            return false;
+        } else if (withdrawMode == WithdrawMode.ANYONE) {
+            return true;
+        } else if (withdrawMode == WithdrawMode.RECIPIENT) {
+            require(withdrawRecipient == account, "WITHDRAW: withdraw only by recipient");
+            return true;
```
Comment (aramalipoor) [inline contracts/common/WithdrawExtension.sol:84]: We need to either "revert" from this method, or return a boolean.

I'd suggest to revert, which means:
- if we want to allow just `return;`
- if we want to check a condition `require(condition, reason)`...
- if we want to revert `revert(reason)`

**Review by aramalipoor — CHANGES_REQUESTED** [review]
To make the assert function to only revert if sender shouldn't withdraw, no return value needed.

**Review by aramalipoor — APPROVED** [review]

**Pull request merged**